python_modules/dagster-graphql/dagster_graphql/schema/asset_graph.py

```diff
@@ -527,28 +528,112 @@ def is_graph_backed_asset(self) -> bool:
-            pass
+            currently_materialized_subset = materialized_partition_subset - failed_partition_subset
+            num_materialized = len(currently_materialized_subset)
+            num_failed = len(failed_partition_subset)
+            if num_materialized == 0 and num_failed == 0:
+                # asset has never been materialized
+                return GrapheneAssetHealthStatus.UNKNOWN
+            if num_failed > 0:
+                return GrapheneAssetHealthStatus.DEGRADED
+            if total_num_partitions > num_failed + num_materialized:
+                # some partitions have never been materialized
+                return GrapheneAssetHealthStatus.WARNING
+            # if no partitions are failed or missing, they must all be successfully materialized
+            return GrapheneAssetHealthStatus.HEALTHY
+
+        asset_record = await AssetRecord.gen(graphene_info.context, asset_key)
+        if asset_record is None:
+            return GrapheneAssetHealthStatus.UNKNOWN
+        asset_entry = asset_record.asset_entry
+
+        if graphene_info.context.instance.can_read_failure_events_for_asset(asset_record):
+            # compute the status based on the asset key table
+            if (
+                asset_entry.last_materialization_storage_id is None
+                and asset_entry.last_failed_to_materialize_storage_id is None
+            ):
+                # never materialized
+                return GrapheneAssetHealthStatus.UNKNOWN
+            if asset_entry.last_failed_to_materialize_storage_id is None:
+                # last_materialization_record must be non-null, therefore the asset successfully materialized
+                return GrapheneAssetHealthStatus.HEALTHY
+            elif asset_entry.last_materialization_storage_id is None:
+                # last_failed_to_materialize_record must be non-null, therefore the asset failed to materialize
+                return GrapheneAssetHealthStatus.DEGRADED
+
+            if (
+                asset_entry.last_materialization_storage_id
+                > asset_entry.last_failed_to_materialize_storage_id
+            ):
+                # latest materialization succeeded
+                return GrapheneAssetHealthStatus.HEALTHY
+            # latest materialization failed
+            return GrapheneAssetHealthStatus.DEGRADED
+        # we are not storing failure events for this asset, so must compute status based on the inforamtion we have available
+        # in some cases this results in reporting as asset as HEALTHY or UNKNOWN during an in progress run
+        # even if the asset was previously failed
         else:
-            # commpute the status based on the asset key table (or using asset status cache for partitioned)
-            pass
+            # if the asset has been successfully materialized in the past, we fallback to that status
```

Comment:
I'm not sure when this code path would be hit. I assume in OSS we won't immediately read failure events for assets, and presumably asset health isn't going to be available for the short term. So will there be any readers to this code path?

Asking because I'm wondering if we need this code path, considering that there are some hairy bits where the status can flip based on edge cases like in progress runs / deleted runs etc.

Reply:
i mainly added it if we want to decouple the observe UI features from enabling the daemon that emits failure events. it'll also be used to determine the status of an asset that doesn't have a failure event written (like if it was last materialized a month ago before the daemon is turned on)

Reply:
Got it. That makes sense